hal/targets/hal/TARGET_ARM_SSG/TARGET_BEETLE/lp_ticker.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * PackageLicenseDeclared: Apache-2.0
+ * Copyright (c) 2015 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cmsis.h"
+#include "device.h"
+#include "objects.h"
+#include "lp_ticker_api.h"
+
+/* Private lp_ticker data */
+/* lp_ticker initialize */
+static uint32_t lp_ticker_initialized = 0;
+/* lp_ticker reload value */
+static uint32_t lp_ticker_reload = 0x0; /* Max Value */
+/* Store Overflow Count */
+static uint32_t lp_ticker_overflows_count = 0;
+
+#if DEVICE_LOWPOWERTIMER
+/**
+ * Interrupt Handler
+ */
+void __lp_ticker_irq_handler(void)
+{
+    if (DualTimer_GetIRQInfo(DUALTIMER0) == SINGLETIMER2) {
+        DualTimer_ClearInterrupt(DUALTIMER0);
+        lp_ticker_overflows_count++;
+    } else {
+        lp_ticker_irq_handler();
+    }
+}
+
+/**
+ * Initialize the low power ticker
+ */
+void lp_ticker_init(void)
+{
+    uint32_t lp_ticker_irqn = 0;
+    /* Verify if lp_ticker has been already Initialized */
+    if (lp_ticker_initialized == 1)
+    {
+        return;
+    }
+    lp_ticker_initialized = 1;
+
+    /* Dualtimer Initialize */
+    DualTimer_Initialize(DUALTIMER0, lp_ticker_reload);
+    /* Dualtimer Enable */
+    DualTimer_Enable(DUALTIMER0, DUALTIMER_COUNT_32
+            //| DUALTIMER_PERIODIC
+            );
+    /* DualTimer get IRQn */
+    lp_ticker_irqn = DualTimer_GetIRQn(DUALTIMER0);
+    /* Enable lp_ticker IRQ */
+    NVIC_SetVector((IRQn_Type)lp_ticker_irqn,
+                (uint32_t)__lp_ticker_irq_handler);
+    NVIC_EnableIRQ((IRQn_Type)lp_ticker_irqn);
+}
+
+/**
+ * Read the current counter
+ * @return: The current timer's counter value in microseconds
+ */
+uint32_t lp_ticker_read(void)
+{
+    uint32_t ticks = 0;
+
+    /* Verify if lp_ticker has not been Initialized */
+    if (lp_ticker_initialized == 0)
+        lp_ticker_init();
+
+    /* Read Timer Value */
+    ticks = DualTimer_Read_2(DUALTIMER0);
+
+    return ticks;
+}
+
+/**
+ * Set interrupt for specified timestamp
+ * timestamp: The time in microseconds to be set
+ */
+void lp_ticker_set_interrupt(timestamp_t timestamp)
+{
+    int32_t delta = 0;
+
+    /* Verify if lp_ticker has been not Initialized */
+    if (lp_ticker_initialized == 0)
+        lp_ticker_init();
+
+    /* Calculate the delta */
+    delta = (int32_t)(timestamp - lp_ticker_read());
+    /* Check if the event was in the past */
+    if (delta <= 0) {
+        /* This event was in the past */
+        lp_ticker_irq_handler();
```

Comment:
Shouldn't this set a flag in the timer to call IRQ and handle this? This could be recursive, that happened before within some targets HAL